docs/workers.md

```diff
@@ -589,15 +589,15 @@ regular expressions:
 
     ^/_matrix/client/(r0|v3|unstable)/keys/upload
 
-If `use_presence` is False in the homeserver config, it can also handle REST
+If `use_presence` is set to `false` in the homeserver config, it can also handle REST
 endpoints matching the following regular expressions:
 
     ^/_matrix/client/(api/v1|r0|v3|unstable)/presence/[^/]+/status
 
 This "stub" presence handler will pass through `GET` request but make the
 `PUT` effectively a no-op.
 
-It will proxy any requests it cannot handle to the main synapse instance. It
+It will proxy not cached requests to the main synapse instance. It
```

Comment:
As of #6964, it looks like the frontend_proxy is really just a generic_worker in disguise. Should we move the frontend_proxy to the "Historical apps" section below?

(I'm very unfamiliar with the history here, maybe @matrix-org/synapse-core  can advise).

Reply:
There are more "Historical apps", IMO.
https://github.com/matrix-org/synapse/blob/6dd7fa12dc18835b18c0632b89c2563d806cd8ef/synapse/app/generic_worker.py#L430-L442

- `synapse.app.media_repository`
- `synapse.app.pusher`
- `synapse.app.federation_sender`

Reply:
Yes indeed, there is absolutely no difference between a `frontend_proxy` worker and a `generic_worker` nowadays; we should just get rid of this section.

... which makes me realise that `generic_worker` can actually handle requests to `^/_matrix/client/(r0|v3|unstable)/keys/upload`, and we should add it to the list up there.

Looks like `^/_matrix/client/(api/v1|r0|v3|unstable)/presence/` is already in the list.

Reply:
I am not sure what I should do? IMO a redesign of the docs should not part of this PR.
Moving URLs to `generic_worker` has impact to docker image. Or deprecate three worker types is a bigger change.

Reply:
The only changes we need (for the documentation) are:
 * remove the section on `frontend_proxy`
 * add `/_matrix/client/(r0|v3|unstable)/keys/upload` to the list of URLs supported by `generic_worker`. It is currently supported, but not documented.

> Moving URLs to generic_worker has impact to docker image.

we may need to update the docker image to match the recommendations from the documentation, but that can be done separately.

Reply:
I do not see this in this PR. This is a bit more.
- `frontend_proxy` needs to be deprecated
- It is not a simple move/add `/_matrix/client/(r0|v3|unstable)/keys/upload`, it needs an extra paragraph for the config param `worker_main_http_uri`, what is needed for the URL

It would make sense to create a PR for depracte all of the redundant workers:
- `synapse.app.media_repository`
- `synapse.app.pusher`
- `synapse.app.federation_sender`

Reply:
> I do not see this in this PR. 

I can't insist, but it just doesn't seem worth making minor corrections to text that is fundamentally outdated.

> This is a bit more.
> 
>     * `frontend_proxy` needs to be deprecated

I don't think we need a particular deprecation notice. `frontend_proxy` is just another example of the redundant workers that are covered under [Historical apps](https://matrix-org.github.io/synapse/latest/workers.html#historical-apps), much like `client_reader` and `event_creator` (which were removed in #7969).

>     * It is not a simple move/add `/_matrix/client/(r0|v3|unstable)/keys/upload`, it needs an extra paragraph for the config param `worker_main_http_uri`, what is needed for the URL

Oh bother, that's true. Yes, we should really move the existing text on `worker_main_http_uri` to the general [Worker configuration](https://matrix-org.github.io/synapse/latest/workers.html#worker-configuration) section. (and probably mention it again next to `/_matrix/client/(r0|v3|unstable)/keys/upload`.

> It would make sense to create a PR for depracte all of the redundant workers:
> 
>     * `synapse.app.media_repository`
>     * `synapse.app.pusher`
>     * `synapse.app.federation_sender`

`media_repository` isn't (yet) redundant, afaik: a `generic_worker` cannot handle the `/media` endpoints. But in general I think it makes more sense to consider each worker app in turn, in separate PRs. #10441 tracks this work.